src/VisualStudio/Core/Test.Next/Services/SolutionAssetCacheTests.cs

```diff
@@ -57,5 +61,57 @@ public async Task TestCleanup()
             // it should not reach here
             Assert.True(false, "asset not cleaned up");
         }
+
+        [Fact]
+        public async Task TestSolutionKeepsAssetPinned()
+        {
+            var workspace = new RemoteWorkspace(FeaturesTestCompositions.RemoteHost.GetHostServices());
+            var solution = workspace.CurrentSolution;
+            var checksums = await solution.State.GetStateChecksumsAsync(CancellationToken.None);
+
+            // Ensure the lazy has computed its value.
+            var storage = new SolutionAssetCache(
+                workspace, cleanupInterval: TimeSpan.FromMilliseconds(1), purgeAfter: TimeSpan.FromMilliseconds(2), gcAfter: TimeSpan.FromMilliseconds(5));
+
+            var checksum1 = checksums.Checksum;
+            var data1 = new object();
+
+            var checksum2 = Checksum.Create(ImmutableArray.CreateRange(Guid.NewGuid().ToByteArray()));
+            var data2 = new object();
+
+            Assert.Equal(data1, storage.GetOrAdd(checksum1, data1));
+            Assert.Equal(data2, storage.GetOrAdd(checksum2, data2));
+
+            var gotChecksum2 = true;
+
+            for (var i = 0; i < 10; i++)
+            {
+                await Task.Delay(10);
+
+                // We should always get this asset, as the solution is keeping it alive.
+                Assert.True(storage.TryGetAsset(checksum1, out object current1));
+                Assert.Equal(data1, current1);
+
+                gotChecksum2 = storage.TryGetAsset(checksum2, out object _);
```

Comment:
Given this is just overwriting it each time, why not just do this outside the loop?